### PR TITLE
usm: add modeler to encoding 

### PR DESCRIPTION
**Pull request by amitslavin**
### What does this PR do?

Add `ConnectionsModeler` struct in order to init encoders once for future batching option, also separate the modeling part from the marshaler.

### Motivation

Separation of concerns between processing the connections and marshaling them into the correct format

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/encoding/encoding.go [hunk 1]
```diff
@@ -58,30 +53,46 @@ func GetUnmarshaler(ctype string) Unmarshaler {
 	return jSerializer
 }
 
-func modelConnections(conns *network.Connections) *model.Connections {
-	cfgOnce.Do(func() {
-		agentCfg = &model.AgentConfiguration{
+type ConnectionsModeler struct {
```
Comment (guyarb) [inline pkg/network/encoding/encoding.go:56]: Document struct and new function

File: pkg/network/encoding/encoding.go [hunk 2]
```diff
@@ -6,10 +6,8 @@
 package encoding
 
 import (
-	"strings"
-	"sync"
-
 	"github.com/gogo/protobuf/jsonpb"
+	"strings"
```
Comment (guyarb) [inline pkg/network/encoding/encoding.go:10]: fix import order

File: pkg/network/encoding/encoding_test.go [hunk 3]
```diff
@@ -328,7 +328,9 @@ func testSerialization(t *testing.T, aggregateByStatusCode bool) {
 		marshaler := GetMarshaler("application/json")
 		assert.Equal("application/json", marshaler.ContentType())
 
-		blob, err := marshaler.Marshal(in)
+		connectionsModeler := InitConnectionsModeler(in)
+		payload := connectionsModeler.ModelConnections(in)
+		blob, err := marshaler.Marshal(payload)
```
Comment (guyarb) [inline pkg/network/encoding/encoding_test.go:333]: For testing purposes, I'd wrap that in a simpler function

File: pkg/network/encoding/json.go [hunk 4]
```diff
@@ -21,11 +19,10 @@ type jsonSerializer struct {
...
 }
 
-func (j jsonSerializer) Marshal(conns *network.Connections) ([]byte, error) {
-	payload := modelConnections(conns)
+func (j jsonSerializer) Marshal(conns *model.Connections) ([]byte, error) {
 	writer := new(bytes.Buffer)
-	err := j.marshaller.Marshal(writer, payload)
-	returnToPool(payload)
+	err := j.marshaller.Marshal(writer, conns)
+	returnToPool(conns)
```
Comment (guyarb) [inline pkg/network/encoding/json.go:25]: should be moved out as well

File: pkg/network/encoding/protobuf.go [hunk 5]
```diff
@@ -8,19 +8,16 @@ package encoding
...
 
 type protoSerializer struct{}
 
-func (protoSerializer) Marshal(conns *network.Connections) ([]byte, error) {
-	payload := modelConnections(conns)
-	buf, err := proto.Marshal(payload)
-	returnToPool(payload)
+func (protoSerializer) Marshal(conns *model.Connections) ([]byte, error) {
+	buf, err := proto.Marshal(conns)
+	returnToPool(conns)
```
Comment (guyarb) [inline pkg/network/encoding/protobuf.go:20]: should be moved out as well

File: cmd/system-probe/modules/network_tracer.go [hunk 6]
```diff
@@ -288,7 +288,9 @@ func getClientID(req *http.Request) string {
 func writeConnections(w http.ResponseWriter, marshaler encoding.Marshaler, cs *network.Connections) {
 	defer network.Reclaim(cs)
 
-	buf, err := marshaler.Marshal(cs)
+	connectionsModeler := encoding.InitConnectionsModeler(cs)
+	payload := connectionsModeler.ModelConnections(cs)
```
Comment (guyarb) [inline cmd/system-probe/modules/network_tracer.go:292]: should add here `returnToPool`

**Review by guyarb — CHANGES_REQUESTED** [review]
CR'ed

File: pkg/network/encoding/protobuf.go [hunk 7]
```diff
@@ -8,19 +8,15 @@ package encoding
...
 const ContentTypeProtobuf = "application/protobuf"
 
 type protoSerializer struct{}
 
-func (protoSerializer) Marshal(conns *network.Connections) ([]byte, error) {
-	payload := modelConnections(conns)
-	buf, err := proto.Marshal(payload)
-	returnToPool(payload)
+func (protoSerializer) Marshal(conns *model.Connections) ([]byte, error) {
+	buf, err := proto.Marshal(conns)
```
Comment (guyarb) [inline pkg/network/encoding/protobuf.go:19]: just return

File: pkg/network/encoding/encoding.go [hunk 8]
```diff
@@ -58,30 +53,48 @@ func GetUnmarshaler(ctype string) Unmarshaler {
...
+	kafkaEncoder                *kafkaEncoder
+	connTelemetryMap            map[string]int64
+	compilationTelemetryByAsset map[string]*model.RuntimeCompilationTelemetry
+	kernelHeaderFetchResult     model.KernelHeaderFetchResult
+	coreTelemetryByAsset        map[string]model.COREResult
+	agentCfg                    *model.AgentConfiguration
+	prebuiltEBPFAssets          []string
+}
+
+func InitConnectionsModeler(conns *network.Connections) *ConnectionsModeler {
```
Comment (guyarb) [inline pkg/network/encoding/encoding.go:69]: document

**Review by guyarb — APPROVED** [review]
LGTM

**Review by hmahmood — APPROVED** [review]

File: pkg/network/encoding/encoding.go [hunk 9]
```diff
@@ -50,6 +50,33 @@ func GetMarshaler(accept string) Marshaler {
 	return jSerializer
 }
 
+// ConnectionsModeler contains all the necessary structs for modeling a connection.
+type ConnectionsModeler struct {
+	httpEncoder  *httpEncoder
+	http2Encoder *http2Encoder
+	kafkaEncoder *kafkaEncoder
+}
```
Comment (guyarb) [inline pkg/network/encoding/encoding.go:58]: Don't we want to add DNS as well?

File: pkg/network/encoding/encoding.go [hunk 10]
```diff
@@ -59,7 +86,7 @@ func GetUnmarshaler(ctype string) Unmarshaler {
 	return jSerializer
 }
 
-func modelConnections(builder *model.ConnectionsBuilder, conns *network.Connections) {
+func modelConnections(builder *model.ConnectionsBuilder, conns *network.Connections, consModeler *ConnectionsModeler) {
```
Comment (guyarb) [inline pkg/network/encoding/encoding.go:89]: I think modelConnections should be a method of ConnectionModeler

File: pkg/network/encoding/json.go [hunk 11]
```diff
@@ -23,8 +23,9 @@ type jsonSerializer struct {
 }
 
 func (j jsonSerializer) Marshal(conns *network.Connections, writer io.Writer) error {
+	connectionsModeler := NewConnectionsModeler(conns)
```
Comment (guyarb) [inline pkg/network/encoding/json.go:26]: don't you need to close the modeler?

File: pkg/network/encoding/protobuf.go [hunk 12]
```diff
@@ -19,8 +19,9 @@ const ContentTypeProtobuf = "application/protobuf"
 type protoSerializer struct{}
 
 func (protoSerializer) Marshal(conns *network.Connections, writer io.Writer) error {
+	connectionsModeler := NewConnectionsModeler(conns)
```
Comment (guyarb) [inline pkg/network/encoding/protobuf.go:22]: don't you need to close the modeler?

File: pkg/network/encoding/json.go (hunk 11 above)
Comment (guyarb) [reply pkg/network/encoding/json.go:26]: the connection modeler should be created once and not everytime we call marshal (preparation for batching)

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/encoding/encoding.go [hunk 13]
```diff
@@ -50,6 +50,37 @@ func GetMarshaler(accept string) Marshaler {
...
+// It also includes formatted connection telemetry related to all batches, not specific batches.
+// Furthermore, it stores the current agent configuration which applies to all instances related to the entire set of connections,
+// rather than just individual batches.
+func NewConnectionsModeler(conns *network.Connections) *ConnectionsModeler {
+	return &ConnectionsModeler{
+		httpEncoder:  newHTTPEncoder(conns.HTTP),
+		http2Encoder: newHTTP2Encoder(conns.HTTP2),
+		kafkaEncoder: newKafkaEncoder(conns.Kafka),
+		ipc:          make(ipCache, len(conns.Conns)/2),
+		dnsFormatter: newDNSFormatter(conns, make(ipCache, len(conns.Conns)/2)),
```
Comment (guyarb) [inline pkg/network/encoding/encoding.go:73]: you need to use `ipc` in `newDNSFormatter`

File: pkg/network/encoding/encoding.go [hunk 14]
```diff
@@ -50,6 +50,37 @@ func GetMarshaler(accept string) Marshaler {
...
+	}
+}
+
+// Close cleans all encoders resources.
```
Comment (guyarb) [inline pkg/network/encoding/encoding.go:77]: what about dns?

**Review by guyarb — COMMENTED** [review]
you need to fix `pkg/network/encoding/encoding_test.go`

File: pkg/network/encoding/encoding.go (hunk 14 above)
Comment (amitslavin) [reply pkg/network/encoding/encoding.go:77]: Actually I do not see any related reset/close functions for the `dnsFormatter` as the decoders have

File: pkg/network/encoding/encoding_test.go [hunk 15]
```diff
@@ -381,15 +384,10 @@ func testSerialization(t *testing.T, aggregateByStatusCode bool) {
 		config.SystemProbe.Set("system_probe_config.collect_dns_domains", false)
 		out := getExpectedConnections(false, httpOutBlob)
 		assert := assert.New(t)
-		marshaler := GetMarshaler("")
```
Comment (guyarb) [inline pkg/network/encoding/encoding_test.go:384]: restore those lines

File: pkg/network/encoding/encoding_test.go [hunk 16]
```diff
@@ -381,15 +384,10 @@ func testSerialization(t *testing.T, aggregateByStatusCode bool) {
...
-		// in case we request empty serialization type, default to application/json
-		assert.Equal("application/json", marshaler.ContentType())
-
-		blobWriter := bytes.NewBuffer(nil)
-		err := marshaler.Marshal(in, blobWriter)
-
 		require.NoError(t, err)
```
Comment (guyarb) [inline pkg/network/encoding/encoding_test.go:387]: remove

File: pkg/network/encoding/encoding_test.go [hunk 17]
```diff
@@ -411,14 +410,7 @@ func testSerialization(t *testing.T, aggregateByStatusCode bool) {
 		out := getExpectedConnections(false, httpOutBlob)
 
 		assert := assert.New(t)
-		marshaler := GetMarshaler("application/whatever")
-
-		// In case we request an unsupported serialization type, we default to application/json
-		assert.Equal("application/json", marshaler.ContentType())
```
Comment (guyarb) [inline pkg/network/encoding/encoding_test.go:417]: restore

**Review by guyarb — APPROVED** [review]

File: pkg/network/encoding/encoding.go [hunk 18]
```diff
@@ -69,20 +101,12 @@ func modelConnections(builder *model.ConnectionsBuilder, conns *network.Connecti
...
-	httpEncoder := newHTTPEncoder(conns.HTTP)
-	defer httpEncoder.Close()
-	kafkaEncoder := newKafkaEncoder(conns.Kafka)
-	defer kafkaEncoder.Close()
-	http2Encoder := newHTTP2Encoder(conns.HTTP2)
-	defer http2Encoder.Close()
 
-	ipc := make(ipCache, len(conns.Conns)/2)
-	dnsFormatter := newDNSFormatter(conns, ipc)
 	tagsSet := network.NewTagsSet()
```
Comment (hmahmood) [inline pkg/network/encoding/encoding.go:105]: `tagSet` is not also a `ConnectionsModeler` level?

Comment (hmahmood) [reply pkg/network/encoding/encoding.go:105]: Same for `routeIndex`

Comment (amitslavin) [reply pkg/network/encoding/encoding.go]: added

**Review by hmahmood — APPROVED** [review]

**Pull request merged**